src/main/java/io/opdev/ExampleValuesReconciler.java

```diff
@@ -90,17 +93,44 @@ private void createFromYaml(String pathToYaml, ExampleValues resource) throws Fi
     log.info("Applying yaml " + pathToYaml + " to the namespace " + resource.getMetadata().getNamespace());
     // Parse a yaml into a list of Kubernetes resources
     List<HasMetadata> result = client.load(new FileInputStream(pathToYaml)).get();
-    for (HasMetadata object : result){
-      ObjectMeta meta = object.getMetadata();
+    for (HasMetadata desiredObject : result){
+      ObjectMeta meta = desiredObject.getMetadata();
       // Patch all objects with owner references
       meta.setOwnerReferences(buildOwnerReference(resource));
-      object.setMetadata(meta);
-      // Apply Kubernetes Resources
-      log.info("Creating resource kind: " + object.getKind() + " with name: " + meta.getName() );
-      client.resource(object).createOrReplace();
+      desiredObject.setMetadata(meta);
+      if (needToUpdateState(desiredObject, resource.getMetadata().getNamespace())){
+         log.info("Creating or updating resource kind: " + desiredObject.getKind() + " with name: " + meta.getName() );
+         client.resource(desiredObject).createOrReplace();
+      }
     }
   }
 
+  private boolean needToUpdateState(HasMetadata desiredObject, String namespace){
+    ResourceDefinitionContext metaContext = new ResourceDefinitionContext.Builder()
+      .withKind(desiredObject.getKind())
+      .withNamespaced(true)
+      .build();
+
+      // Get the existing actual kubernetes resource
+      HasMetadata actualObject = client.genericKubernetesResources(metaContext)
```

Comment:
should probably keep this outside the diff function and pass it the objects to be diffed

Reply:
You're right @skattoju, but the bigger issue is looks like the cast from a `GenericKubernetesResource` to a `HasMetadata` works only in terms of compiler. 
What happens at runtime is at the line `GenericKubernetesResourceMatcher.match()` --> matcher will try parse it for `spec`, and fail because `spec` is actually hidden as a sub-linked-map under the `additionalAttributes` map.
While the desired object is an actual `HasMetadata` and has a spec where the matcher wants it to be.
so I end up comparing between 2 different Java API representations of a k8s resource, and an exception as a result.

Reply:
extracting to method done.